File: services/baidu/asr/baidu_asr_client.cpp
#include "baidu/asr/baidu_asr_client.h"

#include <stdexcept>
#include <vector>

#include <httplib.h>
#include <nlohmann/json.hpp>

#include "async_logger.h"

namespace shizuru::services {

namespace {

// Base64 encoding table.
static const char kBase64Chars[] =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

std::string Base64Encode(const std::string& input) {
  std::string output;
  output.reserve(((input.size() + 2) / 3) * 4);

  const auto* data = reinterpret_cast<const unsigned char*>(input.data());
  size_t len = input.size();

  for (size_t i = 0; i < len; i += 3) {
    unsigned int n = static_cast<unsigned int>(data[i]) << 16;
    if (i + 1 < len) n |= static_cast<unsigned int>(data[i + 1]) << 8;
    if (i + 2 < len) n |= static_cast<unsigned int>(data[i + 2]);

    output.push_back(kBase64Chars[(n >> 18) & 0x3F]);
    output.push_back(kBase64Chars[(n >> 12) & 0x3F]);
    output.push_back((i + 1 < len) ? kBase64Chars[(n >> 6) & 0x3F] : '=');
    output.push_back((i + 2 < len) ? kBase64Chars[n & 0x3F] : '=');
  }

  return output;
}

}  // namespace

BaiduAsrClient::BaiduAsrClient(BaiduConfig config,
                               std::shared_ptr<BaiduTokenManager> token_mgr)
    : config_(std::move(config)), token_mgr_(std::move(token_mgr)) {}

std::string BaiduAsrClient::DetectFormat(const std::string& mime_type) const {
  if (mime_type.find("wav") != std::string::npos) return "wav";
  if (mime_type.find("pcm") != std::string::npos) return "pcm";
  if (mime_type.find("amr") != std::string::npos) return "amr";
  if (mime_type.find("m4a") != std::string::npos) return "m4a";
  return config_.asr_format;
}

std::string BaiduAsrClient::Transcribe(const std::string& audio_data,
                                       const std::string& mime_type) {
  LOG_INFO("[{}] Transcribe audio_len={} mime={}", MODULE_NAME,
           audio_data.size(), mime_type);

  if (audio_data.empty()) {
    LOG_WARN("[{}] Empty audio data", MODULE_NAME);
    return "";
  }

  std::string token = token_mgr_->GetToken();
  std::string format = DetectFormat(mime_type);

  // Build JSON request body.
  nlohmann::json body;
  body["format"] = format;
  body["rate"] = config_.asr_rate;
  body["channel"] = 1;
  body["cuid"] = config_.cuid;
  body["token"] = token;
  body["dev_pid"] = config_.asr_dev_pid;
  body["speech"] = Base64Encode(audio_data);
  body["len"] = static_cast<int>(audio_data.size());

  std::string body_str = body.dump();

  httplib::Client cli(config_.asr_host);
  cli.set_connection_timeout(config_.connect_timeout);
  cli.set_read_timeout(config_.read_timeout);

  auto res = cli.Post(config_.asr_path, body_str, "application/json");

  if (!res) {
    LOG_ERROR("[{}] ASR request failed: no response", MODULE_NAME);
    return "";
  }

  if (res->status != 200) {
    LOG_ERROR("[{}] ASR status {}: {}", MODULE_NAME, res->status, res->body);
    return "";
  }

  LOG_DEBUG("[{}] ASR response: {}", MODULE_NAME, res->body);

  try {
    nlohmann::json j = nlohmann::json::parse(res->body);

    int err_no = j.value("err_no", -1);
    if (err_no != 0) {
      std::string err_msg = j.value("err_msg", "unknown error");
      LOG_ERROR("[{}] ASR error {}: {}", MODULE_NAME, err_no, err_msg);
      return "";
    }

    if (j.contains("result") && j["result"].is_array() &&
        !j["result"].empty()) {
      std::string transcript = j["result"][0].get<std::string>();
      LOG_INFO("[{}] Transcript: \"{}\"", MODULE_NAME, transcript);
      return transcript;
    }

    LOG_WARN("[{}] ASR returned no result", MODULE_NAME);
    return "";
  } catch (const std::exception& e) {
    LOG_ERROR("[{}] Failed to parse ASR response: {}", MODULE_NAME, e.what());
    return "";
  }
}

}  // namespace shizuru::services

File: ui/lib/bridge/baidu_asr_client.dart
// Baidu ASR client — audio to text (Dart port of C++ BaiduAsrClient).

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'baidu_token_manager.dart';

class BaiduAsrClient {
  final BaiduTokenManager tokenManager;
  final String asrHost;
  final String asrPath;
  final String cuid;
  final String defaultFormat;
  final int rate;
  final int devPid;

  BaiduAsrClient({
    required this.tokenManager,
    this.asrHost = 'https://vop.baidu.com',
    this.asrPath = '/server_api',
    this.cuid = 'shizuru_agent',
    this.defaultFormat = 'wav',
    this.rate = 16000,
    this.devPid = 1537, // mandarin + punctuation
  });

  /// Transcribe [audioData] to text. Returns transcript or empty string.
  Future<String> transcribe(Uint8List audioData, {String mimeType = 'audio/wav'}) async {
    if (audioData.isEmpty) return '';

    final token = await tokenManager.getToken();
    final format = _detectFormat(mimeType);
    final speech = base64Encode(audioData);

    final body = jsonEncode({
      'format': format,
      'rate': rate,
      'channel': 1,
      'cuid': cuid,
      'token': token,
      'dev_pid': devPid,
      'speech': speech,
      'len': audioData.length,
    });

    final uri = Uri.parse('$asrHost$asrPath');
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 10);

    try {
      final request = await client.postUrl(uri);
      request.headers.contentType = ContentType.json;
      request.write(body);
      final response = await request.close();

      final responseBody = await response.transform(utf8.decoder).join();

      if (response.statusCode != 200) {
        return '';
      }

      final json = jsonDecode(responseBody) as Map<String, dynamic>;
      final errNo = json['err_no'] as int? ?? -1;
      if (errNo != 0) {
        return '';
      }

      final result = json['result'] as List<dynamic>?;
      if (result != null && result.isNotEmpty) {
        return result[0] as String;
      }
      return '';
    } finally {
      client.close(force: true);
    }
  }

  String _detectFormat(String mimeType) {
    if (mimeType.contains('wav')) return 'wav';
    if (mimeType.contains('pcm')) return 'pcm';
    if (mimeType.contains('amr')) return 'amr';
    if (mimeType.contains('m4a')) return 'm4a';
    return defaultFormat;
  }
}
